File: leetcode/361.bomb-enemy.cpp
// if grid[x] = 'E': f(x) = f(up) + 1
// if grid[x] = 'W': f(x) = 0
// if grid[x] = '0': f(x) = f(up)

class Solution {
public:
    int maxKilledEnemies(vector<vector<char>>& grid) {
        if (grid.size() == 0)
        {
            return 0;
        }
        
        int m = (int)(grid.size());
        int n = (int)(grid[0].size());
        
        vector<vector<int>> direction(m, vector<int>(n, 0));
        vector<vector<int>> table(m, vector<int>(n, 0));
        
        // up
        for(auto i = 0; i < m; ++i)
        {
            for(auto j = 0; j < n; ++j)
            {
                if(i == 0)
                {
                    direction[0][j] = grid[0][j] == 'E' ? 1 : 0;
                }
                else
                {
                    if (grid[i][j] == 'E')
                    {
                        direction[i][j] = direction[i - 1][j] + 1;
                    }
                    else if (grid[i][j] == '0')
                    {
                        direction[i][j] = direction[i - 1][j];
                    }
                    else
                    {
                        direction[i][j] = 0;
                    }
                }
                
                table[i][j] += direction[i][j];
            }
        }
        
        // left
        for(auto i = 0; i < m; ++i)
        {
            for(auto j = 0; j < n; ++j)
            {
                if(j == 0)
                {
                    direction[i][0] = grid[i][0] == 'E' ? 1 : 0;
                }
                else
                {
                    if (grid[i][j] == 'E')
                    {
                        direction[i][j] = direction[i][j - 1] + 1;
                    }
                    else if (grid[i][j] == '0')
                    {
                        direction[i][j] = direction[i][j - 1];
                    }
                    else
                    {
                        direction[i][j] = 0;
                    }
                }
                
                table[i][j] += direction[i][j];
            }
        }
        
        // bottom
        for (auto i = m - 1; i >= 0; --i)
        {
            for(auto j = n - 1; j >=0; --j)
            {
                if(i == m - 1)
                {
                    direction[m - 1][j] = grid[m - 1][j] == 'E' ? 1 : 0;
                }
                else
                {
                    if (grid[i][j] == 'E')
                    {
                        direction[i][j] = direction[i + 1][j] + 1;
                    }
                    else if (grid[i][j] == '0')
                    {
                        direction[i][j] = direction[i + 1][j];
                    }
                    else
                    {
                        direction[i][j] = 0;
                    }
                }
                
                table[i][j] += direction[i][j];
            }
        }
        
        // right
        for (auto i = m - 1; i >= 0; --i)
        {
            for(auto j = n - 1; j >=0; --j)
            {
                if(j == n - 1)
                {
                    direction[i][n - 1] = grid[i][n - 1] == 'E' ? 1 : 0;
                }
                else
                {
                    if (grid[i][j] == 'E')
                    {
                        direction[i][j] = direction[i][j + 1] + 1;
                    }
                    else if (grid[i][j] == '0')
                    {
                        direction[i][j] = direction[i][j + 1];
                    }
                    else
                    {
                        direction[i][j] = 0;
                    }
                }
                
                table[i][j] += direction[i][j];
            }
        }
        
        int res = 0;
        for(auto i = 0; i < m; ++i)
        {
            for(auto j = 0; j < n; ++j)
            {
                if (grid[i][j] == '0')
                {
                    res = max(res, table[i][j]);
                }
            }
        }
        
        return res;
    }
};

File: leetcode/361.bomb-enemy.py
# if grid[x] = 'E': f(x) = f(up) + 1
# if grid[x] = 'W': f(x) = 0
# if grid[x] = '0': f(x) = f(up)

class Solution(object):
    def maxKilledEnemies(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        if grid is None or len(grid) == 0:
            return 0

        m = len(grid)
        n = len(grid[0])

        direction = [[0 for _ in range(n)] for _ in range(m)]
        table = [[0 for _ in range(n)] for _ in range(m)]

        # up
        for i in range(m):
            for j in range(n):
                if i == 0:
                    direction[0][j] = 1 if grid[0][j] == 'E' else 0
                else:
                    if grid[i][j] == 'E':
                        direction[i][j] = direction[i - 1][j] + 1
                    elif grid[i][j] == '0':
                        direction[i][j] = direction[i - 1][j]
                    else:
                        direction[i][j] = 0

                table[i][j] += direction[i][j]

        # left
        for i in range(m):
            for j in range(n):
                if j == 0:
                    direction[i][0] = 1 if grid[i][0] == 'E' else 0
                else:
                    if grid[i][j] == 'E':
                        direction[i][j] = direction[i][j - 1] + 1
                    elif grid[i][j] == '0':
                        direction[i][j] = direction[i][j - 1]
                    else:
                        direction[i][j] = 0

                table[i][j] += direction[i][j]

        # bottom
        for i in range(m - 1, -1, -1):
            for j in range(n - 1, -1, -1):
                if i == m - 1:
                    direction[m - 1][j] = 1 if grid[m - 1][j] == 'E' else 0
                else:
                    if grid[i][j] == 'E':
                        direction[i][j] = direction[i + 1][j] + 1
                    elif grid[i][j] == '0':
                        direction[i][j] = direction[i + 1][j]
                    else:
                        direction[i][j] = 0

                table[i][j] += direction[i][j]

        # right
        for i in range(m - 1, -1, -1):
            for j in range(n - 1, -1, -1):
                if j == n - 1:
                    direction[i][n - 1] = 1 if grid[i][n - 1] == 'E' else 0
                else:
                    if grid[i][j] == 'E':
                        direction[i][j] = direction[i][j + 1] + 1
                    elif grid[i][j] == '0':
                        direction[i][j] = direction[i][j + 1]
                    else:
                        direction[i][j] = 0

                table[i][j] += direction[i][j]

        res = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '0':
                    res = max(res, table[i][j])

        return res
        